libs/native/src/main/java/org/elasticsearch/nativeaccess/lib/NativeLibraryProvider.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.nativeaccess.lib;
+
+import org.elasticsearch.core.internal.provider.ProviderLocator;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+
+/**
+ * Allows loading native library mappings.
+ */
+public abstract class NativeLibraryProvider {
+
+    private final String name;
+    private final Map<Class<?>, Supplier<?>> libraries;
+
+    protected NativeLibraryProvider(String name, Map<Class<?>, Supplier<?>> libraries) {
+        this.name = name;
+        this.libraries = libraries;
+    }
+
+    /**
+     * Get the one and only instance of {@link NativeLibraryProvider} that is specific to the running JDK version.
+     */
+    public static NativeLibraryProvider instance() {
+        return Holder.INSTANCE;
+    }
+
+    /** Returns a human-understandable name for this provider */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * Construct an instance of the given library class.
+     * @param cls The library class to create
+     * @return An instance of the class
+     */
+    public <T> T getLibrary(Class<T> cls) {
+        Supplier<?> libraryCtor = libraries.get(cls);
+        Object library = libraryCtor.get();
+        assert library != null;
+        assert cls.isAssignableFrom(library.getClass());
+        return cls.cast(library);
+    }
+
+    private static NativeLibraryProvider loadProvider() {
+        final int runtimeVersion = Runtime.version().feature();
+        if (runtimeVersion >= 21) {
+            return loadJdkImpl(runtimeVersion);
+        }
+        return loadJnaImpl();
+    }
+
+    private static NativeLibraryProvider loadJdkImpl(int runtimeVersion) {
+        try {
+            var lookup = MethodHandles.lookup();
+            var clazz = lookup.findClass("org.elasticsearch.nativeaccess.jdk.JdkNativeLibraryProvider");
```

Comment:
Is it not possible to use SPI for loading this as well, as we do for the JNA implementation?

Reply:
Unfortunately not. Since the jdk impl is in a separate sourceset, the module-info here in the main sourceset can't know about it. We could in theory have a module-info in the main21 sourceset as well, but then we would need to merge them in the build.

Reply:
Got it, and a JAR cannot define multiple modules. Boo.

The current method just feels a bit odd to me. I don't see why would would use SPI for loading the JNA provider, but reflection for the FFI variant. Do we plan on having further implementations here? Presumably we might if the FFI APIs change in future Java versions, at which point the simple reflection impl won't work. Why aren't we just using the functionality provided by MR jars and have a common class which is simply overridden in each version specific source set?

Reply:
> if the FFI APIs change in future Java versions, at which point the simple reflection impl won't work

This can work mostly as is, with a little refactoring. In the case we need to have eg jdk 22 functionality, we would use a separate, package private class say with a static method doing what we want. Then in main22 we would implement that same class.

> Do we plan on having further implementations here?

No, jna and jdk would be our only two implementations.

> Why aren't we just using the functionality provided by MR jars and have a common class which is simply overridden in each version specific source set?

We could, but I dislike that for two reasons:
(1) The class name would be more confusing, since it would be the same, rather than the implementation specific implementation names
(2) JNA would still need IMPL-JAR to hide the jna dependency, so the version of the class in main would be a pointless wrapper

I understand the dislike for the two different ways of loading these. But I think given there are only 2 implementations, and we are navigating complex classloading/dependency issues, that this is ok. Note that Lucene uses this same manual version checking methodology.

Reply:
> No, jna and jdk would be our only two implementations.

To be specific, there could be multiple JDK version specific implementations, but they would all be loaded the same way, via reflection.

> I understand the dislike for the two different ways of loading these. But I think given there are only 2 implementations, and we are navigating complex classloading/dependency issues, that this is ok.

Fair enough. I thought just using SPI across the board would be an option but yeah, I guess that doesn't work because of the module stuff. Presumably, if you are a module, defining your SPI implementation the "legacy" way via properties file doesn't work?

Reply:
> To be specific, there could be multiple JDK version specific implementations, but they would all be loaded the same way, via reflection.

No, I expect there two be one impl for the jdk. We may isolate and replace bits that have more efficient ways of achieving the intended outcome on newer versions. For example, [this](https://github.com/elastic/elasticsearch/pull/104876/files#diff-8106fe6e4cd808d55cb89e2ebf8e6a67db93bafadddcbdf2526f13747e958307R305) bit to encode a string has a builtin method on Arena in jdk22. I plan to create a utility class with a method to do so in main21, then implement the same class in main22 which uses Arean.allocateFrom. I would not expect to fully implement the native library provider and all its classes in main22 separate from main21. We would utilize the dynamic nature of MR jars to utilize newer functionality.

> Presumably, if you are a module, defining your SPI implementation the "legacy" way via properties file doesn't work?

That's correct, when loading a service provider in a modular setup that we have, the META-INF/services files are not looked at.